src/SectorView.cpp

```diff
@@ -532,6 +549,188 @@ void SectorView::DrawNearSectors(const matrix4x4f& modelview)
 	Gui::Screen::LeaveOrtho();
 }
 
+bool SectorView::MoveRouteItemUp(const std::vector<SystemPath>::size_type element) {
+	if (element <= 0 || element >= m_route.size()) return false;
+
+	SystemPath tmp = m_route[element - 1];
+	m_route[element - 1] = m_route[element];
+	m_route[element] = tmp;
+
+	return true;
+}
+
+bool SectorView::MoveRouteItemDown(const std::vector<SystemPath>::size_type element) {
+	if (element < 0 || element >= m_route.size() - 1) return false;
+
+	SystemPath tmp = m_route[element + 1];
+	m_route[element + 1] = m_route[element];
+	m_route[element] = tmp;
+	return true;
+}
+
+void SectorView::AddToRoute(const SystemPath &path)
+{
+	m_route.push_back(path);
+}
+
+bool SectorView::RemoveRouteItem(const std::vector<SystemPath>::size_type element) {
+	m_route.erase(m_route.begin() + element);
+	return true;
+}
+
+void SectorView::ClearRoute()
+{
+	m_route.clear();
+}
+
+std::vector<SystemPath> SectorView::GetRoute()
+{
+	return m_route;
+}
+
+std::vector<SystemPath> SectorView::AutoRoute(const SystemPath &start, const SystemPath &target)
+{
+	RefCountedPtr<const Sector> start_sec = m_galaxy->GetSector(start);
+	RefCountedPtr<const Sector> target_sec = m_galaxy->GetSector(target);
+
+	// Get the player's hyperdrive from Lua, later used to calculate the duration between systems
+	ScopedTable hyperdrive = ScopedTable(LuaObject<Player>::CallMethod<LuaRef>(Pi::player, "GetEquip", "engine", 1));
+	// Cache max range so it doesn't get recalculated every time we call GetDuration
+	float max_range = hyperdrive.CallMethod<float>("GetMaximumRange", Pi::player);
+	
+	float dist = Sector::DistanceBetween(start_sec, start.systemIndex, target_sec, target.systemIndex);
+
+	int sec_dist = ceilf(dist / Sector::SIZE);
+
+	std::vector<SystemPath> nodes;
+	std::set<std::vector<SystemPath>::size_type> unvisited;
+
+	std::vector<float> path_dist; // distance from source to node
+	std::vector<std::vector<SystemPath>::size_type> path_prev; // previous node in optimal path
+
+	// nodes[0] is always start
+	nodes.push_back(start);
+
+	// go sector by sector for sec_dist sectors and add systems
+	// if they are within 110% of dist of both start and target
```

Comment:
Does this feature allow trying to build a rote to a very remote systems? If it is, it feels like it can be problematic, as memory needed is O(d^3), and runtime is O(d^6), where d is distance between source and destination.

Reply:
I don't see any checks against it. However, I don't see this as a blocking problem, the algorithm can be optimized afterwards *if* this proves problematic, which might not even be the case! ;-)